### PR TITLE
Added comparison counting class, new generators and improved Apostolico-Giancarlo implementation

### PR DESCRIPTION
**Pull request by julianles**

### CODE REVIEW

File: generator/text_pat_generators.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
+import random
+import string
+from scipy.stats import geom
```
Comment (krzysztof-turowski) [inline generator/text_pat_generators.py:3]: Here assumed convention is "import scipy.stats"

File: generator/text_pat_generators.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+class uniform_generator:
+    def generate(text_len, pat_len, alph = string.ascii_lowercase):
```
Comment (krzysztof-turowski) [inline generator/text_pat_generators.py:6]: n, m, A = string.ascii_lowercase

I guess everywhere in the repository A stands for alphabet, t and n for text and its length, w and m for pattern and its length

File: generator/text_pat_generators.py [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+        choice = []
+        for i, val in enumerate(alph):
+            choice += [val for j in range(int(1000*geom.pmf(i+1,p)))]
+        text = ''.join(random.choice(choice) for i in range(text_len))
+        pat = ''.join(random.choice(choice) for i in range(pat_len))
+        return (text, pat)
+
+class natural_generator:
+    data = ""
+    with open('pantadeusz.txt', 'r') as file:
```
Comment (krzysztof-turowski) [inline generator/text_pat_generators.py:22]: Please create input/pantadeusz.txt + even so, please provide a constructor with a parameter "filename", and read data from it

File: benchar/benchar.py [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+        
+        def endswith(self, other):
+            res = str.endswith(self, other)
+            if(len(self) >= len(other)):
+                self._parent_benchar.cmp_count += benchar._first_diff(self[-len(other):], other)
+            return res
+        
+        def __hash__(self):
+            res = str.__hash__(self)
+            return res
```
Comment (krzysztof-turowski) [inline benchar/benchar.py:74]: Isn't the hash function overridden automatically?

File: generator/text_pat_generators.py [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+        text_ind = random.randint(0, len(cls.data)-text_len-1)
+        text = cls.data[text_ind:text_ind+text_len+1]
+        pat_ind = random.randint(0, len(text)-pat_len-1)
+        pat = text[pat_ind:pat_ind+pat_len]
+        return (text, pat)
+
+class bf_hard_generator:
+    def generate(text_len):
+        text = 'a' * text_len + 'b'
+        pat = 'a' * (text_len // 2) + 'b'
```
Comment (krzysztof-turowski) [inline generator/text_pat_generators.py:40]: t, w

File: benchar/benchar.py [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+        def __ge__(self, other):
+            res = str.__ge__(self, other)
+            self._parent_benchar.cmp_count += benchar._first_diff(self, other)
+            return res
+        
+        def endswith(self, other):
+            res = str.endswith(self, other)
+            if(len(self) >= len(other)):
+                self._parent_benchar.cmp_count += benchar._first_diff(self[-len(other):], other)
+            return res
```
Comment (krzysztof-turowski) [inline benchar/benchar.py:70]: Did you test the corner case, when other = '' i.e. self[-len(other):] is the whole string?

File: generator/text_pat_generators.py [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+    def generate(text_len, pat_len, alph = string.ascii_lowercase):
+        text = ''.join(random.choice(alph) for i in range(text_len))
+        pat = ''.join(random.choice(alph) for i in range(pat_len))
+        return (text, pat)
+
+class geometric_generator:
+    def generate(text_len, pat_len, alph = string.ascii_lowercase, p = 0.5):
+        choice = []
+        for i, val in enumerate(alph):
+            choice += [val for j in range(int(1000*geom.pmf(i+1,p)))]
```
Comment (krzysztof-turowski) [inline generator/text_pat_generators.py:15]: Well, shouldn't be that more pythonic?
choice = [v for _ in range(int(1000 * geom.pmf(i+1, p))) for i, v in enumerate(A)]

File: benchar/benchar.py [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+        
+        def __iter__(self):
+            obj = str.__iter__(self)
+            obj = benchar.count_str_iterator(obj, self._parent_benchar)
+            return obj
+        
+        def __eq__(self, other):
+            res = str.__eq__(self, other)
+            self._parent_benchar.cmp_count += benchar._first_diff(self, other)
+            return res
```
Comment (krzysztof-turowski) [inline benchar/benchar.py:39]: If you don't want to get rid of this implementation, please provide another implementation i.e. a class of strings which would (1) count only its own operations, but (2) will implement also reversed ops (e.g. __radd__), and provide fallbacks e.g. in case 'AAAA' == YourString('AAAB') it would store +4 in its internal field of this instance of YourString

To my knowledge 'AAAA' == YourString('AAAB') should throw NotImplemented, and thus [from the documentation] it follows that:
When NotImplemented is returned, the interpreter will then try the reflected operation on the other type, or some other fallback, depending on the operator. If all attempted operations return NotImplemented, the interpreter will raise an appropriate exception.

Therefore, if a.__eq__(b) fails, there should be an invocation of b.__eq__(a).

PS. I know, when t and w are YourStrings, then we need to sum their counters. Or you can add +1 to all arguments that are YourStrings, that's also a valid solution

**Review by krzysztof-turowski — CHANGES_REQUESTED** [review]
Thank you very much for your implementation - overall I'm not convinced with the design of the main class, so I think that at least you should provide its alternative implementation.

**Review by krzysztof-turowski — APPROVED** [review]
Thank you very much, please just fix a remaining print + consider adding tex sources (but overall it's your decision)
